File: Python/Day06.py
import re
import sys

instructions = open('../inputs/Day06.txt').read().split('\n')

lightsOn = 0
brightness = 0

lights = [ [0 for x in range(1000)] for y in range(1000) ]
brights = [ [0 for x in range(1000)] for y in range(1000) ]

for inst in instructions:
	parts = re.search('(?:toggle )?(\w+) (\d+),(\d+) through (\d+),(\d+)', inst).groups()
	action = parts[0]
	x1 = int(parts[1])
	y1 = int(parts[2])
	x2 = int(parts[3])
	y2 = int(parts[4])

	for y in range(y1, y2 + 1):
		for x in range(x1, x2 + 1):
			if action == 'on':
				lights[y][x] = 1
				brights[y][x] += 1
			elif action == 'off':
				lights[y][x] = 0
				brights[y][x] = max(0, brights[y][x] - 1)
			elif action == 'toggle':
				lights[y][x] = abs(lights[y][x] - 1)
				brights[y][x] += 2

lightsOn = sum(map(sum, lights))
brightness = sum(map(sum, brights))

print "Part 1:", lightsOn
print "Part 2:", brightness

File: JavaScript/Day06.js
(function(){
	"use strict";

	require("fs").readFile("../inputs/Day06.txt", "utf8", function(err, data){
		if(err){
			console.log(`Unable to load file: ${err}`);
		} else {

			let lights = [],
				brights = [],
				countLights = (total, light) => total + light;
			for(let y = 0; y < 1000; y++){
				lights[y] = [];
				for(let x = 0; x < 1000; x++){
					lights[y][x] = 0;
				}
			}
			brights = lights.slice().map(row => row.slice());

			let lineParts = /^(?:turn )?(\w*) (\d*?),(\d*?) through (\d*?),(\d*?)$/;

			data.split("\n").map(inst => {
				let parts = inst.match(lineParts),
					action = parts[1],
					x1 = Number(parts[2]),
					y1 = Number(parts[3]),
					x2 = Number(parts[4]),
					y2 = Number(parts[5]);
				for(let y = y1; y <= y2; y++){
					for(let x = x1; x <= x2; x++){
						switch(action){
							case 'on':
								lights[y][x] = 1;
								brights[y][x] += 1;
								break;

							case 'off':
								lights[y][x] = 0;
								brights[y][x] = Math.max(0, brights[y][x] - 1);
								break;

							case 'toggle':
								lights[y][x] = Math.abs(lights[y][x] - 1);
								brights[y][x] = (brights[y][x] || 0) + 2;
						}
					}
				}
			});

			let lightsOn = lights.map(row => row.reduce(countLights)).reduce(countLights);
			let brightness = brights.map(row => row.reduce(countLights)).reduce(countLights);

			console.log(`Part 1: ${lightsOn}`);
			console.log(`Part 2: ${brightness}`);
		}
	});
})();